sdks/python/apache_beam/testing/metric_result_matchers.py

```diff
@@ -0,0 +1,261 @@
+      label_key=typing.Tuple[str, Matcher],
+      label_value=typing.Tuple[str, Matcher],
+      attempted_value=typing.Tuple[int, Matcher],
+      committed_value=typing.Tuple[int, DistributionResult, Matcher])
+  def __init__(self, namespace=None, name=None, step=None, label_key=None,
+               label_value=None, attempted_value=None, committed_value=None):
+    self.namespace = _matcher_or_equal_to(namespace)
+    self.name = _matcher_or_equal_to(name)
+    self.step = _matcher_or_equal_to(step)
+    self.label_key = _matcher_or_equal_to(label_key)
+    self.label_value = _matcher_or_equal_to(label_value)
+    self.attempted_value = _matcher_or_equal_to(attempted_value)
+    self.committed_value = _matcher_or_equal_to(committed_value)
+
+  def _matches(self, metric_result):
+    if self.namespace is not None and not self.namespace.matches(
+        metric_result.key.metric.namespace):
+      return False
+    if self.name and not self.name.matches(metric_result.key.metric.name):
+      return False
+    if self.step and not self.step.matches(metric_result.key.step):
+      return False
+    if (self.attempted_value is not None and
+        not self.attempted_value.matches(metric_result.attempted)):
+      return False
+    if (self.committed_value is not None and
+        not self.committed_value.matches(metric_result.committed)):
+      return False
+    if self.label_key:
+      matched_keys = [key for key in metric_result.key.labels.keys() if
+                      self.label_key.matches(key)]
+      matched_key = matched_keys[0] if matched_keys else None
+      if not matched_key:
+        return False
+      label_value = metric_result.key.labels[matched_key]
+      if self.label_value and not self.label_value.matches(label_value):
+        return False
+    return True
+
+  def describe_to(self, description):
+    if self.namespace:
+      description.append_text(" namespace: ")
+      self.namespace.describe_to(description)
+    if self.name:
+      description.append_text(" name: ")
+      self.name.describe_to(description)
+    if self.step:
+      description.append_text(" step: ")
+      self.step.describe_to(description)
+    if self.label_key:
+      description.append_text(" label_key: ")
+      self.label_key.describe_to(description)
+    if self.label_value:
+      description.append_text(" label_value: ")
+      self.label_value.describe_to(description)
+    if self.attempted_value is not None:
+      description.append_text(" attempted_value: ")
+      self.attempted_value.describe_to(description)
+    if self.committed_value is not None:
+      description.append_text(" committed_value: ")
+      self.committed_value.describe_to(description)
+
+  def describe_mismatch(self, metric_result, mismatch_description):
+    mismatch_description.append_text("was").append_value(metric_result)
+
+
+class DistributionMatcher(BaseMatcher):
+  """A PyHamcrest matcher that validates counter distributions."""
+
+  @with_input_types(
```

Comment:
These decorators are meant to apply to transforms, they have no meaning here. (Also, sum_value, etc. is not a tuple of (int, Matcher).

Reply:
> My first thought when seeing this was asking why we need metrics-specific matchers. If possible, I would make the API for fetching metrics simple enough that one could just use that directly with the standard hamcrest APIs (maybe with a helper or two, such as a DistributionValue matcher).

The main benefit of the matchers, is to allow you to specify these checks, without having to go through the deeper layers of the structure. It also gives us some freedom to change the MetricResult structure, and just change the matcher class. Rather than needing to change all the integration tests which use MetricResult.

Reply:
My original idea was to just populate expected MetricRestuls and compare them. I had done this for MonitoringInfos, but it required writing a lot of code in tests to special case things. Like if you don't want an exact value match. I.e. execution time has an int64_value > 0. With the matchers you can just pass in  greater_than(0) matcher for the value. Then you can reuse the same assertion code in all tests.

We used this technique for the internal metric tests for Dataflow, and it was pretty useful, allowing us to add metric matching to our integration tests without having to write code to query, iterate, etc. over the metrics.

Reply:
But users still have to got through the deeper layers of the structure, and we don't have the freedom to change the MetricsResult due to that, right? In which case it'd be preferable to make the API simple enough that one could write `self.assertGreater(0, some_simple_expression)`. Or perhaps is this really for testing things that are not part of the public API?

(Yes, I'm strongly in favor of being able to use matchers to test things like inequalities.)

Reply:
I agree that we do need to make MetricResult suit our use cases. But I don't want to do that right now. I'm happy to delete these matchers after we iterate on it and this becomes much simpler. 

But the matchers themselves will offer a lot of other benefits anyways, so I still think we should proceed with this.

I do see several problems with MetricResults, which need to be discussed on the dev list:
- MetricResults should probably be a proto, instead of re-implemented in every language.
- MetricResults should not make assumptions like: all metrics have a step associate with them
- MetricResults need to support labels in all languages.
- MetricResults don't have a proper way to reflect how a non-user metric should be returned. I.e. where does the URN go? namespace or name field? etc. Should those details from the MonitoringInfo format even 

Further, as we mentioned in the GetMetrics doc, It will be a very challenging task to upgrade the Dataflow Runner to maintain the URN and all those fields through the entire stack, so MetricResults will have to remain in Beta until at least that design and work is complete. We should not guarantee this format yet either, so I do think it needs to be flexible.

Reply:
I would say it's deeper in that the MetricsResults objects (and constructing a query to get a list of them) is generally overkill for "I set a counter in one of my DoFns, and just want to see its value." But that's also deeper than this change, which seems to have immediate use. I just want to be aware when we're writing adapters because the existing API doesn't suit our needs which solves the problem for us but not all our users.